common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,219 @@
+		messageType := params["message-type"]
+		if messageType != "" {
+			payload.Metadata["messageType"] = []byte(messageType)
+		}
+	case "application/octet-stream":
+		payload.Metadata["encoding"] = []byte("binary/plain")
+	default:
+		return fmt.Errorf("%w: standard content detection failed", errSerializer)
+	}
+	return nil
+}
+
+func isStandardNexusContent(content *nexus.Content) bool {
+	h := content.Header
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		return false
+	}
+	contentType := h.Get("type")
+	if contentType == "" {
+		return len(h) == 0 && len(content.Data) == 0
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return false
+	}
+
+	switch mediaType {
+	case "application/octet-stream",
+		"application/x-temporal-payload":
+		return len(params) == 0
+	case "application/json":
+		if params["format"] == "protobuf" {
+			if params["message-type"] != "" {
+				return len(params) == 2
+			}
+			return len(params) == 1
+		}
+		return len(params) == 0
+	case "application/x-protobuf":
+		if params["message-type"] != "" {
+			return len(params) == 1
+		}
+		return len(params) == 0
+	}
+	return false
+}
+
+// Serialize implements nexus.Serializer.
+func (nexusPayloadSerializer) Serialize(v any) (*nexus.Content, error) {
+	payload, ok := v.(*commonpb.Payload)
+	if !ok {
+		return nil, fmt.Errorf("%w: cannot serialize %v", errSerializer, v)
+	}
+
+	if payload == nil {
+		return &nexus.Content{Header: nexus.Header{}}, nil
+	}
+
+	if !isStandardPayload(payload) {
+		data, err := payload.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("%w: payload marshal error: %w", errSerializer, err)
+		}
+		return &nexus.Content{
+			Header: nexus.Header{"type": "application/x-temporal-payload"},
+			Data:   data,
+		}, nil
+	}
+
+	content := nexus.Content{Header: nexus.Header{}, Data: payload.Data}
+	encoding := string(payload.Metadata["encoding"])
+	messageType := string(payload.Metadata["messageType"])
+
+	switch encoding {
+	case "unknown/nexus-content":
+		for k, v := range payload.Metadata {
+			if k != "encoding" {
+				content.Header[k] = string(v)
+			}
+		}
+	case "json/protobuf":
+		content.Header["type"] = fmt.Sprintf("application/json; format=protobuf")
+		if messageType != "" {
+			content.Header["type"] += fmt.Sprintf("; message-type=%q", messageType)
+		}
+	case "binary/protobuf":
+		content.Header["type"] = fmt.Sprintf("application/x-protobuf")
+		if messageType != "" {
+			content.Header["type"] += fmt.Sprintf("; message-type=%q", messageType)
+		}
+	case "json/plain":
+		content.Header["type"] = "application/json"
+	case "binary/null":
+		// type is unset
+	case "binary/plain":
+		content.Header["type"] = "application/octet-stream"
+	default:
+		return nil, fmt.Errorf("%w: standard payload detection failed", errSerializer)
```

Comment:
Can you report back the invalid encoding so its obvious for users?

Reply:
Yeah, good point.
I originally had a panic here because it's supposed to be unreachable code but changed to an error to be nicer.